tests/unit/controllers/auth.test.ts

```diff
@@ -0,0 +1,179 @@
+import * as jwt from "jsonwebtoken";
+import sinon from "sinon";
+import { User, IUserModel } from "../../../src/database/models/user";
+import { authController } from "../../../src/controllers/auth";
+import { userDBInteractionsStubs, userValidatorStubs, userUtilStubs } from "../stubs/user";
+import { jwtStubs } from "../stubs/auth";
+import { mockReq, mockRes } from "sinon-express-mock";
+import { bcryptPassword } from "../../../src/config/bcrypt";
+import { statusCodes } from "../../../src/config/statusCodes";
+
+let stubs;
+const testUser: IUserModel = new User({
+    username: "test",
+    password: bcryptPassword.generateHash("password"),
+    email: "test@gmail.com",
+    role: "USER",
+    info: {
+        major: "Computer Science",
+        year: "U3",
+        school: "McGill University",
+        bio: "Test account",
+        profilePhoto: "",
+    },
+    problems: [],
+    problemSets: [],
+    platformData: {
+        codeforces: {
+            username: "test",
+            email: "test@gmail.com",
+        },
+    },
+});
+const testToken = jwt.sign({ id: testUser._id, email: testUser.email, role: testUser.role }, process.env.SECRET);
+const testTokenIat = Math.trunc(Date.now() / 1000);
+
+describe("Auth controller tests", () => {
+    before(() => {
+        stubs = {
+            userDB: userDBInteractionsStubs(),
+            userValidator: userValidatorStubs(),
+            userUtil: userUtilStubs(),
+            jwt: jwtStubs()
+        };
+    });
+
+    beforeEach(() => {
+        mockRes.status = sinon.stub().returns(mockRes);
+        mockRes.json = sinon.stub().returns(mockRes);
+    });
+
+    afterEach(() => {
+        sinon.reset();
+    });
+
+    after(() => {
+        stubs.userDB.restore();
+        stubs.userValidator.restore();
+        stubs.userUtil.restore();
+        stubs.jwt.restore();
+    });
+
+    describe("Introspect", () => {
+
+        let req;
+        beforeEach(() => {
+            req = mockReq({
+                query: {
+                    token: testToken
+                }
+            });
+        });
+
+        it("status 200: returns a token with it's user", async () => {
+            stubs.userValidator.validationResult.returns({ isEmpty() { return true; } });
+            await authController.introspect(req, mockRes);
+            sinon.assert.calledWith(mockRes.status, statusCodes.SUCCESS);
+            sinon.assert.calledWith(mockRes.json, { active: true, user: { email: testUser.email, id: `${testUser._id}`, role: testUser.role, iat: testTokenIat } });
+        });
+
+        it("status 422: returns an appropriate response with validation errors", async () => {
```

Comment:
Maybe we can rephrase this as "given validation error" instead of "with validation errors"?